### PR TITLE
[SPARK-36568][SQL] Better FileScan statistics estimation

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR modifies `FileScan.estimateStatistics()` to take the read schema into account.

### Why are the changes needed?
`V2ScanRelationPushDown` can column prune `DataSourceV2ScanRelation`s and change read schema of `Scan` operations. The better statistics returned by `FileScan.estimateStatistics()` can mean better query plans. For example, with this change the broadcast issue in SPARK-36568 can be avoided.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 1]
```diff
@@ -187,7 +189,10 @@ trait FileScan extends Scan
     new Statistics {
       override def sizeInBytes(): OptionalLong = {
         val compressionFactor = sparkSession.sessionState.conf.fileCompressionFactor
-        val size = (compressionFactor * fileIndex.sizeInBytes).toLong
+        val size = (compressionFactor * fileIndex.sizeInBytes /
+          (dataSchema.defaultSize + fileIndex.partitionSchema.defaultSize) *
+          (readDataSchema.defaultSize + readPartitionSchema.defaultSize)).toLong
```
Comment (kevincmchen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:194]: for parquet, which  is a columnar storage format, the  estimated size is much more correct; but for text, csv file,  the size which is computed by the old way, is more correct.

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:194]: Hmm, I don't think that file format has anything to do with this issue. Let me try to explain what I'm trying to fix in this PR. If you take the example V2 example from the description of SPARK-36568 the cost plan looks like this:

```
== Optimized Logical Plan ==
Project [key#251L, col1#252, col2#253, col3#254], Statistics(sizeInBytes=1696.5 GiB)
+- Join Inner, (key#251L = key#259L), Statistics(sizeInBytes=1875.1 GiB)
   :- Filter isnotnull(key#251L), Statistics(sizeInBytes=1385.7 KiB)
   :  +- RelationV2[key#251L, col1#252, col2#253, col3#254] parquet file:/private/var/folders/zb/vz4yf0k97j13m39rsn5s1l5m0000gp/T/spark-9e9cfa69-b858-493a-bbfc-abf83248bf3c/tbl, Statistics(sizeInBytes=1385.7 KiB)
   +- Filter isnotnull(key#259L), Statistics(sizeInBytes=1385.7 KiB)
      +- RelationV2[key#259L] parquet file:/private/var/folders/zb/vz4yf0k97j13m39rsn5s1l5m0000gp/T/spark-9e9cfa69-b858-493a-bbfc-abf83248bf3c/lookup, Statistics(sizeInBytes=1385.7 KiB)
```
The issue here is that `lookup` (2nd `RelationV2`) has the same stats (sizeInBytes=1385.7 KiB) as `tbl` (1st `RelationV2`) does, but we know that `lookup` is column pruned and only returns the `key` column.
Please note that the cost plan is the same if you use other, non-columnar formats.

After this PR the stats will change to:
```
== Optimized Logical Plan ==
Project [key#251L, col1#252, col2#253, col3#254], Statistics(sizeInBytes=199.6 GiB)
+- Join Inner, (key#251L = key#259L), Statistics(sizeInBytes=220.6 GiB)
   :- Filter isnotnull(key#251L), Statistics(sizeInBytes=1385.7 KiB)
   :  +- RelationV2[key#251L, col1#252, col2#253, col3#254] parquet file:/private/var/folders/zb/vz4yf0k97j13m39rsn5s1l5m0000gp/T/spark-be132ca2-1ebb-42a4-b2e1-7518fc0210e8/tbl, Statistics(sizeInBytes=1385.7 KiB)
   +- Filter isnotnull(key#259L), Statistics(sizeInBytes=163.0 KiB)
      +- RelationV2[key#259L] parquet file:/private/var/folders/zb/vz4yf0k97j13m39rsn5s1l5m0000gp/T/spark-be132ca2-1ebb-42a4-b2e1-7518fc0210e8/lookup, Statistics(sizeInBytes=163.0 KiB)
```

**Comment by peter-toth** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/text/TextScanBuilder.scala [hunk 2]
```diff
@@ -33,6 +33,6 @@ case class TextScanBuilder(
   extends FileScanBuilder(sparkSession, fileIndex, dataSchema) {
 
   override def build(): Scan = {
-    TextScan(sparkSession, fileIndex, readDataSchema(), readPartitionSchema(), options)
+    TextScan(sparkSession, fileIndex, schema, readDataSchema(), readPartitionSchema(), options)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/text/TextScanBuilder.scala:36]: shall we pass `dataSchema`?

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala [hunk 3]
```diff
@@ -459,7 +460,9 @@ private[sql] trait SQLTestUtilsBase
    */
   def getLocalDirSize(file: File): Long = {
     assert(file.isDirectory)
-    file.listFiles.filter(f => DataSourceUtils.isDataFile(f.getName)).map(_.length).sum
+    Files.walk(file.toPath).iterator().asScala
+      .filter(p => Files.isRegularFile(p) && DataSourceUtils.isDataFile(p.getFileName.toString))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:464]: what's the actual change? only get regular files?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/text/TextScanBuilder.scala (hunk 2 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/text/TextScanBuilder.scala]: Thanks, fixed in https://github.com/apache/spark/pull/33825/commits/7c606943e81888917988dd252719f06052fc6f56.

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala (hunk 3 above)
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:464]: To support recursive listing for the partitioned table in the added test.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 1 above)
Comment (kevincmchen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:194]: OK, I get it,  thx !!!

**Pull request closed without merging**